R/user-access.R

```diff
@@ -0,0 +1,39 @@
+#' Get information about when users last accessed a dataset
+#'
+#' Useful to get dataset-user interaction metadata about all of the users that
+#' have access to a particular dataset.
+#'
+#' @param dataset_entity a dataset object or a dataset entity (e.g. a single
+#' entry from `selectDatasetCatalog()`)
+#' @param user_fields metadata fields about users to include in the dataframe
+#' (default: c("name", "email", "access_time"))
+#'
+#' @return a dataframe with information about all users for a particular dataset
+#' @export
+userInfo <- function (dataset_entity, user_fields = c("name", "email", "access_time")) {
+    users_catalog <- ShojiCatalog(crGET(paste0(self(dataset_entity),"users/")))
+    all_users <- catalogToDataFrame(users_catalog, keys = user_fields)
+    all_users$dataset <- name(dataset_entity)
+    all_users$dataset_id <- id(dataset_entity)
+
+    return(all_users)
+}
+
+#' Get information about when users last any of the datasets in a project
+#'
+#' Useful to get dataset-user interaction metadata about all of the users that have
+#' access to all of the datasets in a project.
+#'
+#' @param project a character of the project name to gather metadata from
+#' @param ... additional arguments passed to [userInfo()] (e.g. user_fields)
+#'
+#' @return a dataframe with information about all users for a particular dataset
+#' @export
+projectUserInfo <- function (project, ...) {
```

Comment:
Why is this not just `users(project)`? Projects should have the same/similar users catalog API as datasets (and if it doesn't currently, it *should*, so let's do it right--it's on the roadmap).

Reply:
I've added a `users(project)` method that access that. Brian's use case needs the dataset user catalog to get the last_accessed info from, so I've subsumed that under a new `lastAccessed` helper function.

Reply:
So Brian is looking to see the last time a user accessed any dataset within a project? That should be the last_accessed time in the project's user catalog. If it isn't there, we should make it so. Getting every dataset in the project dataset catalog, then getting each dataset's user catalog, isn't the right way to support that.

Reply:
From what I understand, he's actually looking for the last time each user accessed each dataset within a project.

Reply:
Is that a function we need to write? I'm not convinced that that belongs in the library (though I'm open to arguments). And if it does belong as something that we support, then it should be computed server-side and not require N requests to build. 

Seems like if we have natural, good methods, he (and other users) should be able to compose what they want easily, even if we don't deem it a core piece of functionality. Sounds like he wants `lapply(datasets(project), users)`, or some formulation of that, and he can `as.data.frame` the catalogs and `rbind` or whatever, `purrr`ing and piping or however he wants. Perhaps we should do enough make that work (revisiting the `lapply` method definition you mentioned, for example) and leave it at that.

I had planned to make dataset loading be lighter/lazier as a followup to the variable folders work, and it sounds like a use case like this would be another reason to prioritize it.

Reply:
The motivating question is from the YG Ops folks in London. My understanding is that an .xls spreadsheet is as technical as they are willing to go and so I'm trying to automate building a spreadsheet for them to monitor Crunch usage - who looked and last time accessed.

Reply:
Thanks @GitBrianLaw. Can you elaborate on what they actually want to accomplish or know?

Reply:
Sure. I made this by hand https://docs.google.com/spreadsheets/d/1ymk4Pv20iNOranspBP9gXzuA-c9bxBd8FiacOjXV_KM/edit#gid=0
by clicking through the datasets b/c doug said "I need this ASAP" a few weeks ago. Now ops "loves it" and wants it going forward but doing it by hand is crazy. That's when I reached out to Jon. Clear-ish?

Reply:
I totally agree that it's pretty peripheral, but my reasoning for writing and including it was that it would hopefully prevent other implementations that would hit the servers more than necessary (e.g. using `loadDataset()` and then calling `users` on that). I absolutely agree that this is something that we should support server side, which would solve this and a few other requests as well.

From the previous comments, I've got a draft that doesn't over-fetch dataset variables when iterating through a project (though at the cost of having to hack the `users/` endpoint url together). That would solve the immediate need, and we could improve it when we lighten up dataset loading. Additionally, when we get the right user-access history endpoint we could transition that function over to use that without too much disruption. I'm also happy to excise those helpers from the package, and help Brian craft.

Reply:
Yes, thank you. How I'd like to tackle this is (1) enable you to write a simple script to build your xls file whenever by exposing these APIs in R (what I said in https://github.com/Crunch-io/rcrunch/pull/177#discussion_r157080057), and (2) take this and reflect on what kinds of features we should support more fully in terms of usage analytics/client management. I don't think it is the business of the `crunch` R package to generate this bespoke report for you, but it should make it easy for you to assemble the pieces you want. 

Does that make sense?

Reply:
Totally agree and happy to have it implemented that way. Thanks for helping.

Reply:
@jonkeane the question for me is the boundary for what goes in the package (and thus is a feature we own and maintain) vs. what is user's code. That you've written it is great, but I suspect that some of what you written belongs on Brian's computer and not ours. I take the concern about efficiency and silently making it more efficient later, but this specific report looks a little too custom for our general purpose package.

Reply:
I'll take out the helper functions, and send brian the code I have for grabbing what he needs.